src/rgw/rgw_auth_s3.cc

```diff
@@ -239,14 +240,370 @@ bool rgw_create_s3_canonical_header(const req_info& info,
+#define CREDENTIALS_PREFIX_LEN (sizeof(AWS4_HMAC_SHA256_STR) - 1)
+  uint64_t min_len = CREDENTIALS_PREFIX_LEN + 1;
+  if (auth_str.length() < min_len) {
+    dout(10) << "credentials string is too short" << dendl;
+    return -EINVAL;
+  }
+
+  list<string> auth_list;
+  get_str_list(auth_str.substr(min_len), ",", auth_list);
+
+  map<string, string> kv;
+
+  for (string& s : auth_list) {
+    string key, val;
+    int ret = parse_key_value(s, key, val);
+    if (ret < 0) {
+      dout(10) << "NOTICE: failed to parse auth header (s=" << s << ")" << dendl;
+      return -EINVAL;
+    }
+    kv[key] = std::move(val);
+  }
+
+  static std::array<string, 3> aws4_presigned_required_keys = {
+    "Credential",
+    "SignedHeaders",
+    "Signature"
+  };
+
+  for (string& k : aws4_presigned_required_keys) {
+    if (kv.find(k) == kv.end()) {
+      dout(10) << "NOTICE: auth header missing key: " << k << dendl;
+      return -EINVAL;
+    }
+  }
+
+  credential = std::move(kv["Credential"]);
+  signedheaders = std::move(kv["SignedHeaders"]);
+  signature = std::move(kv["Signature"]);
+
+  /* sig hex str */
+  dout(10) << "v4 signature format = " << signature << dendl;
+
+  /* ------------------------- handle x-amz-date header */
+
+  /* grab date */
+
+  const char *d = info.env->get("HTTP_X_AMZ_DATE");
+  struct tm t;
+  if (!parse_iso8601(d, &t, NULL, false)) {
+    dout(10) << "error reading date via http_x_amz_date" << dendl;
+    return -EACCES;
+  }
+  date = d;
+
+  return 0;
+}
+
+int parse_credentials(const req_info& info,             /* in */
+                      std::string& access_key_id,       /* out */
+                      std::string& credential_scope,    /* out */
+                      std::string& signedheaders,       /* out */
+                      std::string& signature,           /* out */
+                      std::string& date,                /* out */
+                      bool& using_qs)                   /* out */
+{
+  const char* const http_auth = info.env->get("HTTP_AUTHORIZATION");
+  using_qs = http_auth == nullptr || http_auth[0] == '\0';
+
+  int ret;
+  std::string credential;
+  if (using_qs) {
+    ret = parse_v4_credentials_qs(info, credential, signedheaders,
+                                  signature, date);
+  } else {
+    ret = parse_v4_credentials_hdrs(info, credential, signedheaders,
+                                    signature, date);
+  }
+
+  if (ret < 0) {
+    return ret;
+  }
+
+  /* AKIAIVKTAZLOCF43WNQD/AAAAMMDD/region/host/aws4_request */
+  dout(10) << "v4 credential format = " << credential << dendl;
+
+  if (std::count(credential.begin(), credential.end(), '/') != 4) {
+    return -EINVAL;
+  }
+
+  /* credential must end with 'aws4_request' */
+  if (credential.find("aws4_request") == std::string::npos) {
+    return -EINVAL;
+  }
+
+  /* grab access key id */
+  const size_t pos = credential.find("/");
+  access_key_id = credential.substr(0, pos);
```

Comment:
DONE. Switched to `boost::string_view` for both `access_key_id` and `credential` as well.